### ISSUE TITLE
Merge authservice-go into the main project as a v1 version

### ISSUE THREAD

**Issue by nacx** [istio-ecosystem/authservice#242]
There has been an effort to improve code readability, testability, quality, and the overall maintainability of the project. The current C++ codebase is hard to maintain and to properly test end-to-end, and adding features like better integration in Kubernetes is hard. To that end, the [authservice-go](https://github.com/tetrateio/authservice-go) is a full rewrite of `authservice` in Go, making the project more accessible to the cloud-native community and easier to integrate into its ecosystem.

It is a drop-in replacement, fully compatible with the existing image. The project contains [e2e test suites](https://github.com/tetrateio/authservice-go/blob/main/DEVELOPMENT.md#backward-compatibility-tests) that run with both images, the `authservice-go` one and the C++ `authservice` to guarantee it s a clean drop-in replacement.

The rewrite adds [extensive code coverage](https://app.codecov.io/gh/tetrateio/authservice-go), several [end-to-end test suites](https://github.com/tetrateio/authservice-go/tree/main/e2e) that verify the correct behavior of the supported OIDC flows, as well as the mentioned [compatibility suite](https://github.com/tetrateio/authservice-go/tree/main/e2e/legacy).

In addition to the extensive tests and feature parity, it also fixes the following issues:

* https://github.com/istio-ecosystem/authservice/issues/56
* https://github.com/istio-ecosystem/authservice/issues/59
* https://github.com/istio-ecosystem/authservice/issues/61
* https://github.com/istio-ecosystem/authservice/issues/101
* https://github.com/istio-ecosystem/authservice/issues/151
* https://github.com/istio-ecosystem/authservice/issues/234

---

I'd like to propose merging `authservice-go` into the upstream `authservice` and to use the Go-based version as the main option moving forward, as it is much better tested and usable. I propose to create a `v0` branch to keep the C++ code should anyone be interested in continuing that, and use the main branch for `authservice-go` and call it the `v1` version. The reason for having the Go version in the main branch is to properly accommodate the Go module mechanism (having the `go.mod` file in the main repo branch) and properly convey the project's path forward.

@ZackButcher @zinuga I'd love to hear your feedback on this!

**Comment by dio** [comment]
Sounds good!

**Comment by nacx** [comment]
Thanks!

The merge has been done:

* I've created a `v0-c++` branch to keep the old code should anyone want to continue maintaining and contributing to that version.
* I've also kept the `master` branch with the old code to avoid breaking existing open PRs.
* I've created the `main` branch with the Go rewrite, and configured the repo to use `main` as the default branch from now on.
* Created an initial v1.0.0 release based on the new Go authservice.
